artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java

```diff
@@ -318,28 +397,178 @@ public void postAcknowledge(MessageReference ref, AckReason reason) throws Excep
+      Message messageCommand = createMessage(ref.getQueue().getAddress(), ref.getQueue().getName(), POST_ACK, nodeID, internalID, reason);
+      if (sync) {
+         OperationContext operationContext;
+         operationContext = OperationContextImpl.getContext(server.getExecutorFactory());
+         messageCommand.setUserContext(OperationContext.class, operationContext);
+         if (tx == null) {
+            // notice that if transactional, the context is lined up on beforeCommit as part of the transaction operation
+            operationContext.replicationLineUp();
+         }
+      }
+
+      if (tx != null) {
+         MirrorACKOperation operation = getAckOperation(tx);
+         // notice the operationContext.replicationLineUp is done on beforeCommit as part of the TX
+         operation.addMessage(messageCommand, ref);
+      } else {
+         server.getStorageManager().afterStoreOperations(new IOCallback() {
+            @Override
+            public void done() {
+               try {
+                  logger.debug("preAcknowledge::afterStoreOperation for messageReference {}", ref);
+                  route(server, messageCommand);
+               } catch (Exception e) {
+                  logger.warn(e.getMessage(), e);
+               }
+            }
+
+            @Override
+            public void onError(int errorCode, String errorMessage) {
+            }
+         });
+      }
+   }
+
+   private MirrorACKOperation getAckOperation(Transaction tx) {
+      MirrorACKOperation ackOperation = (MirrorACKOperation) tx.getProperty(TransactionPropertyIndexes.MIRROR_ACK_OPERATION);
+      if (ackOperation == null) {
+         logger.trace("getAckOperation::setting operation on transaction {}", tx);
+         ackOperation = new MirrorACKOperation(server);
+         tx.putProperty(TransactionPropertyIndexes.MIRROR_ACK_OPERATION, ackOperation);
+         tx.afterStore(ackOperation);
+      }
+
+      return ackOperation;
+   }
+
+   private MirrorSendOperation getSendOperation(Transaction tx) {
+      if (tx == null) {
+         return null;
+      }
+      MirrorSendOperation sendOperation = (MirrorSendOperation) tx.getProperty(TransactionPropertyIndexes.MIRROR_SEND_OPERATION);
+      if (sendOperation == null) {
+         logger.trace("getSendOperation::setting operation on transaction {}", tx);
+         sendOperation = new MirrorSendOperation();
+         tx.putProperty(TransactionPropertyIndexes.MIRROR_SEND_OPERATION, sendOperation);
+         tx.afterStore(sendOperation);
+      }
+
+      return sendOperation;
+   }
+
+   private static class MirrorACKOperation extends TransactionOperationAbstract {
+
+      final ActiveMQServer server;
+
+      // This pair contains the Message used to generate the command towards the target, the reference being acked
+      final List<Pair<Message, MessageReference>> acks = new ArrayList<>();
```

Comment:
There's no remove... The add is always single threaded.

Each Message has its own Reference. I could just use an user property on the message.. but it would have the same effect honestly.

Reply:
I don't think it will make much difference, but I will use setUseContext on the message command stored in the Operation.

Reply:
So, no removals  as the entire thing is discarded along with the parent tx object, ok. (Oops, I hadnt got to looking at the rest when commenting that, didnt go back hehe),

If the 'UserContext' from the message is already by something else, then there likely isnt much difference as you say, though it will need to do a lookup (I was thinking a map here, it could foreach). If it isnt already in use, then starting to use it would be worse than the original Pair's.

So yeah...maybe just put back the annoying Pair. It only applies in the Sync case, right?

Reply:
I had made a comment last week and I pressed start a review by accident (weird), and all my comments were lost.. 

I actually mentioned I didn't think it would make much difference... and I just realized what you meant on the Map.

I'm using a Map on a separate commit. If you could take a look and I will squash it after you look.

I don't think it will make much difference on the runtime... although the code is cleaner this way.

Reply:
@gemmellr I will just squash the separate commit. It looks better actually.